### PR TITLE
Added energy margin capability to Fe recommendation scripts

### LINKED ISSUE #2745: Provide acceptable energy loss margin capability for CA recommendation scripts
> estimate 1
Parent [feature/story]: #2744

The implementation for the Fe recommendation feature was modified because it did not allow the user to specify an energy margin (additional energy that can be spent to pick a lower frequency value for Fe).  This is beneficial when noisy data (due to a low volume of runs or system characteristics) is provided to the parsing script and guidance from the end user can help provide a more reasonable characterization.

Done criteria:

- [x] Integration tests that cover the feature run and pass
- [x] Dependencies mocked in unit tests are still valid
- [x] Documentation changes are required since this change does
      not modify interfaces or features.

### PR DESCRIPTION
**Pull request by lhlawson**
Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

- Relates to #2744 story from github issues
- Fixes #2745 change request from github issues.
- Relies on #2722 & #2723

The Fe recommendation scripts have been updated to provide --<core/uncore/gpu>-energy-margin options that allow the user to specify what percentage of additional energy consumption is acceptable when attempting to find an Fe value lower than the absolute minimum energy point. 

This capability helps to address the issue of noisy data when characterizing a system.

This PR refactors some common code from the Fe recommendation scripts into experiment.util and adds usage of the frequency step (in dev) and gpu frequency step (PR #2722 & #2723) to the scripts.  

Updates
- [x] cpu Fe recommendation script has --core & --uncore energy margin options & warnings
- [x] gpu Fe recommendation script has energy margin options

### CODE REVIEW

File: service/docs/source/geopm_agent_cpu_activity.7.rst [hunk 1]
```diff
@@ -146,6 +146,17 @@ CPU compute activity agent ConstConfigIOGroup configuration file can then be gen
 
     integration/experiment/uncore_frequency_sweep/gen_cpu_activity_constconfig_recommendation.py --path <UNCORE_SWEEP_DIR> --region-list "intensity_1","intensity_16"
 
+Depending on the number of runs, system noise, and other factors there may be more than one reasonable
+value for ``Fe`` for a given domain.  In these cases a warning similar to the following will be provided::
+
+    'Warning: Found N possible alternate Fe value(s) within 5% energy consumption of Fe for <Control>.
+     Consider using the energy-margin options.\n'
+
+If the occurs the user may chose to use the provided configuration file OR rerun the recommendation script with
```
Comment (dannosliwcd) [inline service/docs/source/geopm_agent_cpu_activity.7.rst]:
```suggestion
If the occurs the user may choose to use the provided configuration file OR rerun the recommendation script with
```

File: service/docs/source/geopm_agent_cpu_activity.7.rst [hunk 2]
```diff
@@ -146,6 +146,17 @@ CPU compute activity agent ConstConfigIOGroup configuration file can then be gen
...
+any the energy-margin options ``--core-energy-margin`` & ``--uncore-energy-margin`` along with a value such
+as 0.05 (5%). These options will attempt to identify a lower ``Fe`` for the respective domain that costs less than
+the energy consumed at ``Fe`` plus the energy-margin percentage provided.
```
Comment (dannosliwcd) [inline service/docs/source/geopm_agent_cpu_activity.7.rst:158]: What happens if just outside of the users's 5% margin also shows a lot of nearby samples? Should the user stay at 5%, increase the margin, suspect something is wrong, or something else?

File: service/docs/source/geopm_agent_gpu_activity.7.rst [hunk 3]
```diff
@@ -80,11 +80,22 @@ The ``Phi`` input is the only policy value.
...
     integration/experiment/gpu_frequency_sweep/gen_gpu_activity_constconfig_recommendation.py --path <GPU_SWEEP_DIR>
 
-This version of the agent uses ConstConfigIO to provide per node Fe values.
+Depending on the number of runs, system noise, and other factors there may be more than one reasonable
+value for ``Fe``.  In these cases a warning similar to the following will be provided::
+
+    'Warning: Found N possible alternate Fe value(s) within 5% energy consumption of Fe for <frequency>.
+     Consider using the energy-margin options.\n'
+
+If the occurs the user may chose to use the provided configuration file or rerun the recommendation script with
```
Comment (dannosliwcd) [inline service/docs/source/geopm_agent_gpu_activity.7.rst]:
```suggestion
If the occurs the user may choose to use the provided configuration file or rerun the recommendation script with
```

File: integration/experiment/util.py [hunk 4]
```diff
@@ -166,3 +166,43 @@ def merge_const_config(new_config, old_config_path=None):
...
+    if len(df_mean) > 1:
+        if energy_margin != 0.0:
+            sys.stderr.write('Found Fe = {} with energy = {}.  Searching for alternate '
+                             'based on an energy margin of {}\n'.format(energy_efficient_frequency,min_energy, energy_margin))
+
+            # Grab all energy readings associated with frequencies that a 1Hz below Fe
+            # TODO: Consider iloc instead and just grab all idx prior to Fe
+            df_mean = df_mean.loc[:energy_efficient_frequency - 1]
+
+            # Find any energy reading that is within 5% of Fe's energy while having a lower frequency
```
Comment (dannosliwcd) [inline integration/experiment/util.py]:
```suggestion
            # Find any energy reading that is within a given margin of Fe's energy while having a lower frequency
```

File: integration/experiment/util.py [hunk 5]
```diff
@@ -166,3 +166,43 @@ def merge_const_config(new_config, old_config_path=None):
...
+            energy_in_margin = [e for e in df_mean if (e - min_energy) / e < energy_margin]
+
+            # If any values are within margin
+            if energy_in_margin:
+                min_energy = max(energy_in_margin);
```
Comment (dannosliwcd) [inline integration/experiment/util.py]:
```suggestion
                min_energy = min(energy_in_margin);
```
I think?

File: integration/experiment/util.py [hunk 6]
```diff
@@ -166,3 +166,43 @@ def merge_const_config(new_config, old_config_path=None):
...
+                # Store the associated frequency
+                energy_efficient_frequency = df_mean[df_mean == min_energy].index[0];
```
Comment (dannosliwcd) [inline integration/experiment/util.py]:
```suggestion
                energy_efficient_frequency = df_mean.loc[df_mean == min_energy].index[0];
```
Or  idxmin() might work.

File: service/docs/source/geopm_agent_cpu_activity.7.rst (hunk 2 above)
Comment (lhlawson) [reply service/docs/source/geopm_agent_cpu_activity.7.rst:158]: In general when there is a significant amount of clustering values in testing it's been related to bad or noisy data.  Additional trials can help address this, as can checking the system configuration (are features disabled, is there some background processes skewing data, etc).  

In other cases we would expect the user to evaluate the data based upon their usage of the CA recommendations, which may include increasing the error of margin they use & exploring the impact on their workloads.

File: integration/experiment/util.py [hunk 7]
```diff
@@ -174,34 +174,38 @@ def energy_efficient_frequency(df, freq_col_name, energy_col_name, energy_margin
...
-            energy_in_margin = [e for e in df_mean if (e - min_energy) / e < energy_margin]
+            # Find any energy reading that is within a given margin of Fe's energy while having a lower frequency
+            energy_in_margin = [e for e in df_mean if (e - energy_reading) / e < energy_margin]
 
             # If any values are within margin
             if energy_in_margin:
-                min_energy = max(energy_in_margin);
+                # Since we've been given a margin, we're going to find the maximum energy within
+                # the margin as that is generally associated with lower frequencies
+                energy_reading = max(energy_in_margin);
```
Comment (dannosliwcd) [inline integration/experiment/util.py]: Looks like the goal is to get `min(frequency near_energy(5%) argmin_freq(energy))`, but I think we're computing  `argmax_freq(energy near_energy(5%) min(energy))` as a proxy instead. Why use a proxy computation when the desired result requires similar operations over the same data?

File: integration/experiment/util.py [hunk 8]
```diff
@@ -187,16 +187,17 @@ def energy_efficient_frequency(df, freq_col_name, energy_col_name, energy_margin
             df_mean = df_mean.loc[:energy_efficient_frequency - 1]
 
             # Find any energy reading that is within a given margin of Fe's energy while having a lower frequency
-            energy_in_margin = [e for e in df_mean if (e - energy_reading) / e < energy_margin]
+            energy_max_acceptable = e = energy_reading + energy_reading * energy_margin;
```
Comment (dannosliwcd) [inline integration/experiment/util.py]:
```suggestion
            energy_max_acceptable = energy_reading + energy_reading * energy_margin;
```

**Review by dannosliwcd — DISMISSED** [review]

File: integration/experiment/util.py [hunk 9]
```diff
@@ -166,3 +166,48 @@ def merge_const_config(new_config, old_config_path=None):
...
+                             'based on an energy margin of {}\n'.format(energy_efficient_frequency,
+                                                                        energy_reading, energy_margin))
+
+            # Grab all energy readings associated with frequencies that a 1Hz below Fe
+            # TODO: Consider iloc instead and just grab all idx prior to Fe
+            df_mean = df_mean.loc[:energy_efficient_frequency - 1]
+
+            # Find any energy reading that is within a given margin of Fe's energy while having a lower frequency
+            energy_max_acceptable = energy_reading + energy_reading * energy_margin;
+            df_in_margin = df_mean.loc[df_mean <= energy_max_acceptable];
```
Comment (bgeltz) [inline integration/experiment/util.py]:
```suggestion
            energy_max_acceptable = energy_reading + energy_reading * energy_margin
            df_in_margin = df_mean.loc[df_mean <= energy_max_acceptable]
```

I see come C syntax sneaking in. :)

File: service/docs/source/geopm_agent_gpu_activity.7.rst [hunk 10]
```diff
@@ -80,11 +80,22 @@ The ``Phi`` input is the only policy value.
 ConstConfigIOGroup Configuration File Generation
 ------------------------------------------------
 
+This version of the agent uses ConstConfigIO to provide per node Fe values.
```
Comment (bgeltz) [inline service/docs/source/geopm_agent_gpu_activity.7.rst]:
```suggestion
This version of the agent uses ConstConfigIO to provide per-node Fe values.
```

**Review by bgeltz — CHANGES_REQUESTED** [review]

File: integration/experiment/gpu_frequency_sweep/gen_gpu_activity_constconfig_recommendation.py [hunk 11]
```diff
@@ -83,6 +84,10 @@ def get_config_from_frequency_sweep(full_df):
     parser = argparse.ArgumentParser()
     parser.add_argument('--const-config-path', required=False, default=None,
                         help='path containing existing ConstConfigIO configuration file')
+    parser.add_argument('--gpu-energy-margin', default=0, type=float, dest='gpu_energy_margin',
+                        help='Percentage of additional energy it is acceptable to consume if it results '
+                             'in a lower frequency selection for Fe.  This is useful for analyzing '
+                             'noisy systems that have many gpu frequencies near the Fe energy consumption value')
```
Comment (bgeltz) [inline integration/experiment/gpu_frequency_sweep/gen_gpu_activity_constconfig_recommendation.py]:
```suggestion
                             'noisy systems that have many GPU frequencies near the Fe energy consumption value')
```

File: integration/experiment/uncore_frequency_sweep/gen_cpu_activity_constconfig_recommendation.py [hunk 12]
```diff
@@ -90,46 +88,53 @@ def frequency_recommendation(df_region_group, region, domain):
...
+    # Round to step size
+    uncore_freq_recommendation = round(uncore_freq_recommendation / frequency_step) * frequency_step
 
-    # Then analyze the core senstivite region (intensity_16)
+    # Then analyze the core senstivite region (ex: intensity_16)
     # to find the most efficient core frequency
     # when running at the uncore_freq_efficient determined
     # above
     df = df[df['uncore-frequency (Hz)'] == uncore_freq_recommendation]
+
```
Comment (bgeltz) [inline integration/experiment/uncore_frequency_sweep/gen_cpu_activity_constconfig_recommendation.py]:
```suggestion
```

File: integration/experiment/machine.py [hunk 13]
```diff
@@ -113,6 +116,13 @@ def _query(self):
         for sn in signal_names:
             self.signals[sn] = util.geopmread('{} board 0'.format(sn))
 
+        conditional_signal_names = ['GPU_CORE_FREQUENCY_STEP']
+        for sn in signal_names:
```
Comment (bgeltz) [inline integration/experiment/machine.py]:
```suggestion
        for sn in conditional_signal_names:
```

**Review by bgeltz — CHANGES_REQUESTED** [review]

File: service/docs/source/geopm_agent_cpu_activity.7.rst [hunk 14]
```diff
@@ -146,6 +146,17 @@ CPU compute activity agent ConstConfigIOGroup configuration file can then be gen
 
     integration/experiment/uncore_frequency_sweep/gen_cpu_activity_constconfig_recommendation.py --path <UNCORE_SWEEP_DIR> --region-list "intensity_1","intensity_16"
 
+Depending on the number of runs, system noise, and other factors there may be more than one reasonable
+value for ``Fe`` for a given domain.  In these cases a warning similar to the following will be provided::
+
+    'Warning: Found N possible alternate Fe value(s) within 5% energy consumption of Fe for <Control>.
+     Consider using the energy-margin options.\n'
+
+If the occurs the user may choose to use the provided configuration file OR rerun the recommendation script with
```
Comment (avilcheslopez) [inline service/docs/source/geopm_agent_cpu_activity.7.rst]:
```suggestion
If this occurs the user may choose to use the provided configuration file OR rerun the recommendation script with
```

File: integration/experiment/gpu_frequency_sweep/gen_gpu_activity_constconfig_recommendation.py [hunk 15]
```diff
@@ -94,7 +99,12 @@ def get_config_from_frequency_sweep(full_df):
                          '; run a frequency sweep before using this analysis.\n')
         sys.exit(1)
 
-    output = get_config_from_frequency_sweep(df)
+    if args.gpu_energy_margin < 0:
+        sys.stderr.write('<geopm> Error: GPU energy margin must be non-negative\n')
```
Comment (avilcheslopez) [inline integration/experiment/gpu_frequency_sweep/gen_gpu_activity_constconfig_recommendation.py]: When printing warnings, I've seen we place the "<geom>" tag after the "Warning" header/label. I do see that for printing errors, the pattern is to place it before. Is there a reason why we place the "<geopm>" differently for errors and warnings?

File: integration/experiment/machine.py [hunk 16]
```diff
@@ -113,6 +116,13 @@ def _query(self):
...
+            try:
+                self.signals[sn] = util.geopmread('{} board 0'.format(sn))
+            except:
+                sys.stderr.write('Skipping conditional signal: {}'.format(sn))
```
Comment (avilcheslopez) [inline integration/experiment/machine.py]: Does this need a "Warning" label?

File: integration/experiment/util.py [hunk 17]
```diff
@@ -166,3 +166,48 @@ def merge_const_config(new_config, old_config_path=None):
...
+    Find the frequency that provides the minimum package energy consumption
+    within the dataframe provided.
+    """
+    df_mean = df.groupby(freq_col_name)[energy_col_name].mean()
+    energy_efficient_frequency = df_mean.idxmin()
+    energy_reading = df_mean[energy_efficient_frequency];
+
+    if len(df_mean) > 1:
+        if energy_margin != 0.0:
+            sys.stderr.write('Found Fe = {} with energy = {}.  Searching for alternate '
```
Comment (avilcheslopez) [inline integration/experiment/util.py]: Warning/Error label + ``<geopm>`` tag?

File: integration/experiment/util.py [hunk 18]
```diff
@@ -166,3 +166,48 @@ def merge_const_config(new_config, old_config_path=None):
...
+
+            # If any values are within margin
+            if len(df_in_margin) > 1:
+                # Since we've been given a margin, we're going to find the minimum frequency within
+                # that margin
+                energy_efficient_frequency = df_in_margin.index.min()
+                # Associated energy reading
+                energy_reading = df_in_margin.loc[energy_efficient_frequency];
+
+                sys.stderr.write('Found alternate Fe = {} with energy = {}.\n'.format(energy_efficient_frequency,
```
Comment (avilcheslopez) [inline integration/experiment/util.py]: Warning/Error label + ``<geopm>`` tag?

File: integration/experiment/util.py [hunk 19]
```diff
@@ -166,3 +166,48 @@ def merge_const_config(new_config, old_config_path=None):
...
+                                                                                      energy_reading))
+            else:
+                sys.stderr.write('No alternate found with energy-margin {}.  Using Fe={}.\n'.format(energy_margin,
```
Comment (avilcheslopez) [inline integration/experiment/util.py]: Warning/Error label + ``<geopm>`` tag?

File: integration/experiment/util.py [hunk 20]
```diff
@@ -166,3 +166,48 @@ def merge_const_config(new_config, old_config_path=None):
...
+                                                                                                    energy_efficient_frequency))
+
+        else:
+            df_mean = df_mean.loc[:energy_efficient_frequency - 1]
+            nearby_energy_count = len([e for e in df_mean if (e - energy_reading) / e < 0.05]);
+            sys.stderr.write('Warning: Found {} possible alternate Fe value(s) within 5% '
```
Comment (avilcheslopez) [inline integration/experiment/util.py]: ``<geopm>`` tag?

File: integration/experiment/gpu_frequency_sweep/gen_gpu_activity_constconfig_recommendation.py (hunk 15 above)
Comment (lhlawson) [reply integration/experiment/gpu_frequency_sweep/gen_gpu_activity_constconfig_recommendation.py]: From review: Error/Warning: <geopm> script_name: Actual message"

**Review by avilcheslopez — APPROVED** [review]

**Review by bgeltz — APPROVED** [review]

**Pull request merged**